File: java-code/src/main/java/refactoring/organizing_data/change_reference_to_value/after/Person.java
package refactoring.organizing_data.change_reference_to_value.after;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

/**
 * After step3 our telephoneNumber object is immutable
 */
@Setter
@Getter
@AllArgsConstructor
public class Person {
  private TelephoneNumber telephoneNumber;

  public String getOfficeAreaCode() {
    return this.telephoneNumber.getAreaCode();
  }

  public void setOfficeAreaCode(String areaCode) {
    //STEP 2: Create new Object for each setting
    this.telephoneNumber = new TelephoneNumber(areaCode, this.telephoneNumber.getNumber());

  }

  public String getOfficeNumber() {
    return this.telephoneNumber.getNumber();
  }

  public void setOfficeNumber(String number) {
    //STEP 3: Create new Object for each setting
    this.telephoneNumber = new TelephoneNumber(this.telephoneNumber.getAreaCode(), number);
  }


}

File: go-code/organizing_data/change_reference_to_value/after/person.go
package after

type Person struct {
	telephoneNumber TelephoneNumber
}

func (p *Person) getOfficeAreaCode() string {
	return p.telephoneNumber.areaCode
}

// STEP 2: Create new Object for each setting
func (p *Person) setOfficeAreaCode(areaCode string) {
	p.telephoneNumber = New(areaCode, p.telephoneNumber.getNumber())
}

func (p *Person) getOfficeNumber() string {
	return p.telephoneNumber.getNumber()
}

// STEP 3: Create new Object for each setting
func (p *Person) setOfficeNumber(officeNumber string) {
	p.telephoneNumber = New(p.telephoneNumber.getAreaCode(), officeNumber)
}
